internal/amqp/publisher.go

```diff
@@ -0,0 +1,100 @@
+package amqp
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/lunarway/release-manager/internal/log"
+	"github.com/lunarway/release-manager/internal/tracing"
+	"github.com/pkg/errors"
+	"github.com/streadway/amqp"
+)
+
+type rawPublisher struct {
+	channel      *amqp.Channel
+	exchangeName string
+	routingKey   string
+}
+
+func newPublisher(amqpConn *amqp.Connection, exchangeName string) (publisher, error) {
+	channel, err := amqpConn.Channel()
+	if err != nil {
+		return nil, errors.WithMessage(err, "get channel")
+	}
+	err = declareExchange(channel, exchangeName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "declare exchange")
+	}
+	return &rawPublisher{
+		channel:      channel,
+		exchangeName: exchangeName,
+	}, nil
+}
+
+func (p *rawPublisher) Publish(ctx context.Context, eventType, messageID string, message interface{}) error {
+	data, err := json.Marshal(message)
+	if err != nil {
+		return errors.WithMessage(err, "marshal message")
+	}
+	pub := amqp.Publishing{
+		ContentType:   "application/json",
+		Type:          eventType,
+		Body:          data,
+		MessageId:     messageID,
+		CorrelationId: tracing.RequestIDFromContext(ctx),
+	}
+	// TODO: add publisher confirms to the channel and wait for acknowledgement
+	// before returning
+	err = p.channel.Publish(p.exchangeName, p.routingKey, false, false, pub)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *rawPublisher) Close() error {
+	err := p.channel.Close()
+	if err != nil {
+		return errors.WithMessage(err, "close amqp channel")
+	}
+	return nil
+}
+
+type loggingPublisher struct {
```

Comment:
Move out of the package as it is not relevant for the amqp package.

Reply:
You'll also wait with moving this one out?